### ISSUE TITLE
Increase accuracy of torso parsing

### ISSUE THREAD

**Issue by RobinRenggli** [YudongGuo/AD-NeRF#9]
(Firstly, I want to apologize for creating multiple issues; if this is bothersome or you are very busy just let me know and I will try to work these out myself. In this case, feel free to close the issue.)

I've tried several videos now, and in many cases the parsing is not very accurate for the torso, especially when the subject is a woman with long hair. The problem is, if the torso isn't correctly parsed, it influences the generated background image.

Are there parameters in the parsing model that can be adjusted, or is it just not perfect regarding the torso? Do you have any recommendations for what to change if the results of step 3 in the preprocessing aren't good?

Here are some examples:
![1](https://user-images.githubusercontent.com/32697212/124189280-c34b4c80-dac0-11eb-8a97-db85ed5033ca.png)
![253](https://user-images.githubusercontent.com/32697212/124189300-ca725a80-dac0-11eb-8f4a-3ea733b0c5a0.png)
![7011](https://user-images.githubusercontent.com/32697212/124189390-ed9d0a00-dac0-11eb-9c4c-1c2d56b7fdbd.png)

**Comment by YudongGuo** [comment]
Hi, feel free to create issues. 

For face pasing, we just use the open source code [face-parsing.PyTorch](https://github.com/zllrunning/face-parsing.PyTorch), and I don't think there are adjustable parameters to improve the results. In my opinion, the most promising way to improve the results is to use more challenging dataset to retrain a more powerful model.

 I notice there are some released face parsing dataset [lapa-dataset](https://github.com/JDAI-CV/lapa-dataset) and [short-video-face-parsing-challenge](https://maadaa.ai/cvpr2021-short-video-face-parsing-challenge/), and including these data for training the parsing model may improve the parsing accuracy.

**Comment by saslamsameja** [comment]
@RobinRenggli have you faced similar issues with male subjects as well?

**Comment by RobinRenggli** [comment]
> 
> 
> @RobinRenggli have you faced similar issues with male subjects as well?

I did not run many tests on male subjects, but the factors that seem to confuse the parsing models are what you'd expect: Anything that obstructs the torso, such as hair or for example a microphone in front of the person. With "easy" examples it works well.

**Comment by RobinRenggli** [comment]
I've played around with this a little and managed to improve it, but not by much and only for one specific example (so the opposite of generalizable). In the process I started to ask myself the following question:

Where exactly is the torso parsing used, except for creating the com_imgs and the background? Why are the com_imgs used for the loss in the torso-nerf, when, if the parsing and background creation works well, com_imgs should be near identical to orig_imgs? I tried using the orig_imgs for the loss instead and it seems to work similarly. Is it because the slight differences in background between com_imgs and orig_imgs leads to artifacts and slows down/confuses the torso learning sometimes? 

Improving this seems very useful, because right now the set of input videos that produce good results are somewhat constrained by this (short haired person facing the camera and not moving too much gets a little boring). Especially since the pose generation seems to work for these harder examples, but the parsing fails.

**Comment by YudongGuo** [comment]
> I've played around with this a little and managed to improve it, but not by much and only for one specific example (so the opposite of generalizable). In the process I started to ask myself the following question:
> 
> Where exactly is the torso parsing used, except for creating the com_imgs and the background? Why are the com_imgs used for the loss in the torso-nerf, when, if the parsing and background creation works well, com_imgs should be near identical to orig_imgs? I tried using the orig_imgs for the loss instead and it seems to work similarly. Is it because the slight differences in background between com_imgs and orig_imgs leads to artifacts and slows down/confuses the torso learning sometimes?
> 
> Improving this seems very useful, because right now the set of input videos that produce good results are somewhat constrained by this (short haired person facing the camera and not moving too much gets a little boring). Especially since the pose generation seems to work for these harder examples, but the parsing fails.

Hi, thanks for sharing this. Sometimes (specifically, when the background does not change) using ori_imgs will improve the result. Since in com_imgs we only copy the pixels coverd by head and torso parsing maps from ori_imgs, com_imgs will not contain the full torso part if the torso paring map fails to cover the full torso. If the background doesn't change during the full video, the perfect com_imgs ('perfect' means extracted from perfect parsing maps) will be the same as the ori_imgs. The reason that we use com_imgs is that sometimes the background changes.

So the conclusion is, if the head parsing maps are good while the torso parsing maps are bad and the background dosen't show obvious changes, using ori_imgs will lead better results than com_imgs.

**Comment by RobinRenggli** [comment]
Thank you for the elaborate answer. I can confirm that your conclusion worked for me as well.

I've been running an experiment over the weekend on a "difficult" example to test the limits of your method. (Difficult to me means long hair, non-white background, a lot of head-movement and with that torso movement.)
I found that the main issue with these examples is the torso, if the person does not stand very still, the Torso-NeRF produces bad outputs for me. 
Have you tried similar examples and did you get similar results? Did you test any ideas on how to handle this, or is the model just not suited to examples where the person does not stand very still (i.e. news anchor style)?

**Comment by YudongGuo** [comment]
Hi, I have tested on a video in which the person kept moving head and torso significantly, the synthesized torso parts show obvious artifacts and are often blur. So it's a limitation of the current method.
